<a href="https://colab.research.google.com/github/Konstantin5054232/ausbildungsprojekte/blob/main/02_Kreditnehmer_Zuverl%C3%A4ssigkeitsstudie/Kreditnehmer_Zuverl%C3%A4ssigkeitsstudie.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Kreditnehmer-Zuverlässigkeitsstudie**

Die Kreditabteilung der Bank muss folgende Fragen beantworten: 
* Gibt es einen Zusammenhang zwischen Kinderwunsch und fristgerechter Kreditrückzahlung? 
* Gibt es einen Zusammenhang zwischen Familienstand und pünktlicher Kreditrückzahlung? 
* Gibt es einen Zusammenhang zwischen Einkommen und fristgerechter Kreditrückzahlung? 
* Wie wirken sich unterschiedliche Verwendungszwecke des Darlehens auf die fristgerechte Rückzahlung aus?

Die Bank erhielt Statistiken über die Zahlungsfähigkeit der Kunden.

Die Ergebnisse der Studie werden beim Aufbau eines Kredit-Scoring-Modells angewendet.

# Datenexploration

In [1]:
# Wir importieren die notwendigen Bibliotheken
import pandas as pd

In [2]:
# Wir laden Tabellen mit Daten
data = pd.read_csv('/content/loaners.csv')

In [3]:
# Wir werden die erhaltenen Daten studieren
display(data.head())
display(data.tail())
display(data.info())
display(data.shape)

for column in data.columns:
    display(data[column].value_counts())

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

children  days_employed  dob_years education  education_id  \
21520         1   -4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1   -2113.346888         38   среднее             1   
21523         3   -3112.481705         38   среднее             1   
21524         2   -1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

        total_income                       purpose  
21520  224791.862382             операции с жильем  
21521  155999.806512          сделка с автомобилем  
21522   89672.561153                  недвижимость  
21523  244093.050500  на покупку своего автомобиля  
21524   82047.418899         на покупку автомобиля

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


None

(21525, 12)

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

-8437.673028      1
-3507.818775      1
 354500.415854    1
-769.717438       1
-3963.590317      1
                 ..
-1099.957609      1
-209.984794       1
 398099.392433    1
-1271.038880      1
-1984.507589      1
Name: days_employed, Length: 19351, dtype: int64

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
22    183
66    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      668
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

женат / замужем          12380
гражданский брак          4177
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

сотрудник          11119
компаньон           5085
пенсионер           3856
госслужащий         1459
безработный            2
предприниматель        2
студент                1
в декрете              1
Name: income_type, dtype: int64

0    19784
1     1741
Name: debt, dtype: int64

253875.639453    1
157691.851056    1
70113.902786     1
116196.518662    1
157205.834562    1
                ..
168880.592469    1
148042.721049    1
60039.334460     1
175979.762960    1
82047.418899     1
Name: total_income, Length: 19351, dtype: int64

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
покупка жилья для сдачи                   653
операции с жильем                         653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      624
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**Schlussfolgerungen**

Es gibt Fehler in den Daten, höchstwahrscheinlich aufgrund menschlicher Fehler oder Upload-Fehler. Die Spalte 'children' enthält 47 Mal den Wert „-1“ und 76 Mal den Wert „20“, und es gibt keine Kreditnehmer mit 6 bis 19 Kindern.Die Spalte 'days_employed' enthält nur negative Werte oder Artefakte: 400281 Tage , 340266 Tage (1361 Jahr), auch wenn es Arbeitsstunden sind, sind es ungefähr 170 Jahre. Der älteste Kreditnehmer ist 75 Jahre alt, sodass die maximale Anzahl an Arbeitstagen etwa 15.000 betragen kann: (75-18) * 250. Die Spalte „dob_years“ enthält außerdem 101 „0“-Werte. Die Spalte 'education' enthält Werte in Groß- und Kleinschreibung. Die Spalte 'gender' enthält einen XNA-Wert. Die Spalte 'purpose' enthält viele ähnliche Zwecke. Alle Spalten außer 'days_employed' und 'total_income'  enthalten 21525 Nicht-Null-Werte, und diese beiden Spalten enthalten die gleiche Anzahl von Werten, nämlich 19351. Die Null-Werte in diesen Spalten stimmen höchstwahrscheinlich zeilenübergreifend überein. Dies wird durch die Zeilen 12 und 21510 bestätigt.

# Datenvorverarbeitung

## Umgang mit fehlenden Daten

In [4]:
# Wir sehen uns eine Tabelle an, die Zeilen mit Nullen in den Spalten 'days_employed' und 
# 'total_income' enthält und wie viele es sind.
display(data[data['days_employed'].isnull()].head())
display(data[data['days_employed'].isnull()].count())
display(data[data['total_income'].isnull()].head())
display(data[data['total_income'].isnull()].count())

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   

    total_income                           purpose  
12           NaN                   сыграть свадьбу  
26           NaN                       образование  
29           NaN  строительство жилой недвижимости  
41           NaN  сделка с подержанным автомобилем  
55           NaN                   сыграть свадьбу

children            2174
days_employed          0
dob_years           2174
education           2174
education_id        2174
family_status       2174
family_status_id    2174
gender              2174
income_type         2174
debt                2174
total_income           0
purpose             2174
dtype: int64

In [6]:
# Wir berechnen das arithmetische Mittel und den Median für die angegebenen Spalten.
print('days_employed, mean', data['days_employed'].mean())
print('days_employed, median', data['days_employed'].median())
print('total_income, mean', data['total_income'].mean())
print('total_income, median', data['total_income'].median())

days_employed, mean 63046.49766147338
days_employed, median -1203.369528770489
total_income, mean 167422.30220817294
total_income, median 145017.93753253992


Der arithmetische Mittelwert für die Spalte 'days_employed' beträgt 63047 Tage, also über 200 Jahre. In diesem Fall sind die Werte in der Spalte entweder negativ oder unrealistische Werte. Verwenden Sie den Median, um Nullwerte einzugeben.

In [7]:
# Wir werden Nullwerte mit der gewählten Methode füllen.
median_days_employed = data['days_employed'].median()
data['days_employed'] = data['days_employed'].fillna(median_days_employed)
# Wir werden prüfen, ob in der Tabelle noch Nullwerte vorhanden sind.
data[data['days_employed'].isnull()].count()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

Für die Spalte „total_income“ unterscheiden sich das arithmetische Mittel und die Medianwerte leicht, daher verwenden wir auch den Median, um Nullwerte einzugeben.

In [8]:
# Wir durchlaufen die eindeutigen „income_type“-Werte, berechnen den Median 
# für jede Gruppe und füllen die Lücken aus.
for income_type in data['income_type'].unique():
    median = data.loc[data['income_type'] == income_type, 'total_income'].median()
    data.loc[(data['income_type'] == income_type) & (data['total_income'].isna()), 'total_income'] = median
# Wir prüfen, ob die Tabelle Nullwerte enthält.
data[data['total_income'].isnull()].count()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

In [9]:
# Wir füllen die Lücken in der Spalte total_income auf andere Weise aus.
data['total_income'] = data.groupby('income_type')['total_income'].apply(lambda x: x.fillna(x.median()))
# Wir prüfen, ob die Tabelle Nullwerte enthält.
data[data['total_income'].isnull()].count()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

**Schlussfolgerungen**

Die Spalten „Tagesbeschäftigung“ und „Gesamteinkommen“ enthielten jeweils 2174 Nullwerte. Mögliche Gründe für das Auftreten von Datenlücken können die Zurückhaltung von Klienten sein, Angaben zu ihrem Einkommen und ihrer Dienstzeit zu machen, oder ihre Abwesenheit (z. B. bei Studenten). Sie sollten diese Zeilen nicht löschen, weil das sind 10,1 % aller Daten (2174/21525). Jetzt gibt es keine Nullwerte in der Tabelle.

## Ersetzen des Datentyps

In [10]:
# Wir werden die Werte der Spalten, die Informationen über die Dienstzeit der Kreditnehmer und 
# ihr Einkommen enthalten, in ganze Zahlen umwandeln.
data['days_employed'] = data['days_employed'].astype('int')
data['total_income'] = data['total_income'].astype('int')
# Mal sehen was passiert.
data.head()

children  days_employed  dob_years education  education_id  \
0         1          -8437         42    высшее             0   
1         1          -4024         36   среднее             1   
2         0          -5623         33   Среднее             1   
3         3          -4124         32   среднее             1   
4         0         340266         53   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

**Schlussfolgerungen**

Die Gesamtarbeitserfahrung wird in ganzen Tagen gemessen, daher müssen die Werte in der Spalte 'days_employed' in Ganzzahlen umgewandelt werden. Die Werte der Spalte 'total_income' werden ebenfalls in ganze Zahlen umgewandelt, weil eine solche Genauigkeit ist nicht erforderlich, um das Problem zu lösen. Einkommensbruchteile können sich in der Tabelle ergeben haben, indem das Jahreseinkommen des Kreditnehmers durch 12 Monate geteilt wurde. Bruchteile von Erfahrungswerten, gemessen in Jahren und Monaten, könnten durch Division durch die Anzahl der Tage in einem Jahr oder in einem Monat entstehen.

## Doppelte Handhabung

In [11]:
# Wir zählen die Anzahl der Duplikate
data.duplicated().sum()

54

In [12]:
# Wir werden die Spalte 'education' klein schreiben und die Anzahl der Duplikate erneut zählen.
data['education'] = data['education'].str.lower()
data.duplicated().sum()

71

In [13]:
# Das Ergebnis ist 71. Es gibt also Zeilen, die sich nur in Großbuchstaben in der Spalte 'education' unterscheiden. 
# Sie müssen auch entfernt werden.
data = data.drop_duplicates().reset_index(drop = True)

In [14]:
# Wir prüfen die Anzahl der Duplikate und Zeilen in der Tabelle.
display(data.duplicated().sum())
data.shape

0

(21454, 12)

**Schlussfolgerungen**

Es gibt 0 Duplikate, die Anzahl der Zeilen in der Tabelle beträgt 21454, d.h. 71 weniger. Durch Mitarbeiterfehler können Dubletten entstanden sein. Beispielsweise hat ein Mitarbeiter Informationen über denselben Kunden zweimal eingegeben, verschiedene Mitarbeiter haben Informationen über denselben Kunden eingegeben und so weiter.

## Lemmatisierung

In [20]:
# Wir werden die Lemmatisierungsbibliothek laden.
from pymystem3 import Mystem
m = Mystem()

In [21]:
# Wir erstellen eine Liste der eindeutigen Zwecke der Spalte 'purpose' und untersuchen sie.
purpose_list = data['purpose'].unique()
purpose_list

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

In [22]:
# Wir lemmatisieren jedes der Ziele.
lemmas = []
for row in purpose_list:
    lemma = m.lemmatize(row)
    lemmas.append(lemma)
# Wir werden das Ergebnis sehen und studieren.
lemmas

KeyboardInterrupt: ignored

Es gibt offensichtliche Ausgabengruppen für den Kauf eines Autos, eine Ausbildung, die Organisation einer Hochzeit. Es ist nicht möglich, den Zweck eines Darlehens zu trennen, um ein Haus für sich selbst zu kaufen oder ein Haus zu vermieten, weil Es gibt einen Zweck einfach "Gehäuse". Es ist auch nicht möglich, den Wohnungskauf vom Erwerb einer Gewerbeimmobilie zu trennen, weil Es gibt ein Ziel "Immobilien", ohne seinen Zweck zu spezifizieren. Deshalb vereinen wir diese Ziele zu einem: „Immobilien“.

In [ ]:
# Wir werden eine Funktion erstellen, um den Zweck der Kreditaufnahme zu standardisieren.
def purpose_standart(row):
    purpose = m.lemmatize(row['purpose'])
    if 'автомобиль' in purpose:
        return 'автомобиль'
    if 'свадьба' in purpose:
        return 'свадьба'
    if 'образование' in purpose:
        return 'образование'
    if 'недвиж' or 'жил' in purpose:
        return 'недвижимость'

    # Wir werden eine Spalte mit standardisierten Zielen an die Tabelle anhängen.
    data['purpose_standart'] = data.apply(purpose_standart, axis=1) 
    data.head()

**Schlussfolgerungen**

Informationen über Kreditnehmer in die Datenbank wurden höchstwahrscheinlich von verschiedenen Bankangestellten eingegeben, die keine klaren Anweisungen hatten, wie der Zweck der Kreditaufnahme widergespiegelt werden sollte. Wir haben alle Ziele auf eine Einheitlichkeit gebracht, indem wir 4 Gruppen gebildet haben. Diese Informationen werden nun weiter analysiert.

## Datenkategorisierung

In [ ]:
# Zur weiteren Analyse teilen wir Kreditnehmer in folgende Gruppen ein: 
# „keine Kinder“ und „haben Kinder“.
# Wir schreiben eine Funktion zur Aufteilung in Gruppen.
def children_class(row):
    children = row['children']
    debt = row['debt']
    
    if children == 0:
        return 'keine Kinder'
                
    if children != 0:
        return 'es gibt Kinder'

In [ ]:
# Wir bilden eine zusätzliche Spalte zur Tabelle und sehen uns diese an.
data['children_class'] = data.apply(children_class, axis=1)
display(data.head())

In [ ]:
# Wir werden die Kunden der Bank nach Einkommen in 3 Gruppen einteilen: 
# bis 100 (niedrig), von 100 bis 250 (mittel), über 250 (hoch).
# Wir schreiben eine Funktion zur Aufteilung in Gruppen.
def income_class(row):
    total_income = row['total_income']
    debt = row['debt']
    if total_income < 100000:
        return 'niedriges Einkommen'
        
    if 100000 <= total_income < 250000:
        return 'mittelis Einkommen'
        
    if total_income >= 250000:
        return 'hohes Einkommen'

In [ ]:
# Wir bilden eine zusätzliche Spalte zur Tabelle und sehen uns diese an.
data['income_class'] = data.apply(income_class, axis=1)
data.head()

**Schlussfolgerungen**

Wir haben alle Kunden in Gruppen eingeteilt. Die erhaltenen Informationen ermöglichen es uns, die vom Kunden gestellten Fragen zu beantworten.